File: src/schema/merchant.rs
use std::path::Display;
use crate::schema::service::Service;
use bson::oid::ObjectId;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Default)]
pub struct Merchant {
    #[serde(rename = "_id")]
    pub id: ObjectId,
    pub name: String,
    pub description: Option<String>,
    pub chain: Option<String>, // Name of the chain if part of a chain store series
    pub entity: ObjectId,      // Reference to the Entity
    pub beacon_code: String, // Unique identifier for the merchant for displaying in the beacon name
    pub area: ObjectId,
    pub r#type: MerchantType,
    /// Hex color code for UI representation,
    /// e.g., `#00704a` for Starbucks green,
    /// whereas `#ffc72c` for McDonald's yellow.
    pub color: Option<String>,
    /// List of tags for categorization, e.g., "food", "electronics", "clothing"
    /// Tags can be used for search and filtering
    pub tags: Vec<String>,
    pub location: (f64, f64),
    pub style: MerchantStyle,
    pub polygon: Option<Vec<(f64, f64)>>, // List of (x, y) pairs of coordinates
    pub available_period: Option<Vec<(i64, i64)>>, // List of (start_time, end_time) in milliseconds on a 24-hour clock
    pub email: Option<String>,
    pub phone: Option<String>,
    pub website: Option<String>,
    pub social_media: Option<Vec<SocialMedia>>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Default)]
#[serde(rename_all = "kebab-case")]
pub enum MerchantType {
    Food {
        /// Type of cuisine, e.g., `Italian`, `Chinese`.
        cuisine: Option<FoodCuisine>,
        /// Type of food establishment, e.g., `Restaurant`, `Cafe`.
        r#type: FoodType,
    },
    Electronics {
        mobile: bool,      // Whether it sells in mobile devices
        computer: bool,    // Whether it sells in computers
        accessories: bool, // Whether it sells accessories
    },
    Clothing {
        menswear: bool,      // Whether it's menswear
        womenswear: bool,    // Whether it's womenswear
        childrenswear: bool, // Whether it's childrenswear
    },
    Supermarket,
    Health,
    Entertainment,
    Facility {
        /// Type of facility, e.g., `Restroom`, `ATM`.
        r#type: FacilityType,
    },
    /// The room is, for example, a hotel room, office room, or meeting room.
    /// It may or may not use authentication or access control, but it only has several doors
    /// that can be used to enter or exit the room.
    Room,
    #[default]
    Other, // For any other type not listed
}

impl std::fmt::Display for MerchantType {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            MerchantType::Food { cuisine, r#type } => {
                if let Some(cuisine) = cuisine {
                    write!(f, "{:?} {:?}", r#type, cuisine)
                } else {
                    write!(f, "{:?}", r#type)
                }
            }
            MerchantType::Electronics {
                mobile,
                computer,
                accessories,
            } => {
                let mut types = vec![];
                if *mobile {
                    types.push("Mobile");
                }
                if *computer {
                    types.push("Computer");
                }
                if *accessories {
                    types.push("Accessories");
                }
                write!(f, "Electronics ({})", types.join(", "))
            }
            MerchantType::Clothing {
                menswear,
                womenswear,
                childrenswear,
            } => {
                let mut types = vec![];
                if *menswear {
                    types.push("Menswear");
                }
                if *womenswear {
                    types.push("Womenswear");
                }
                if *childrenswear {
                    types.push("Childrenswear");
                }
                write!(f, "Clothing ({})", types.join(", "))
            }
            MerchantType::Supermarket => write!(f, "Supermarket"),
            MerchantType::Health => write!(f, "Health"),
            MerchantType::Entertainment => write!(f, "Entertainment"),
            MerchantType::Facility { r#type } => write!(f, "Facility ({:?})", r#type),
            MerchantType::Room => write!(f, "Room"),
            MerchantType::Other => write!(f, "Other"),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Default)]
#[serde(rename_all = "kebab-case")]
pub enum FacilityType {
    Restroom,
    Atm,
    InformationDesk,
    #[default]
    Other, // For any other type not listed
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Default)]
#[serde(rename_all = "kebab-case")]
pub enum FoodType {
    Restaurant(FoodCuisine),
    Cafe,
    FastFood,
    Bakery,
    Bar,
    #[default]
    Other, // For any other type not listed
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "kebab-case")]
pub enum FoodCuisine {
    Italian,
    Chinese {
        cuisine: ChineseFoodCuisine, // Specific type of Chinese cuisine
        specific: Option<String>,    // Specific dish or style, e.g., "Dim Sum", "Ningbo Cuisine"
    },
    Indian,
    American,
    Japanese,
    Korean,
    French,
    Thai,
    Mexican,
    Mediterranean,
    Spanish,
    Vietnamese,
    MiddleEastern,
    African,
    Other(String), // For any other type not listed
}

impl Default for FoodCuisine {
    fn default() -> Self {
        FoodCuisine::Other("Unknown".to_string())
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Copy, Default)]
#[serde(rename_all = "kebab-case")]
pub enum ChineseFoodCuisine {
    Cantonese,
    Sichuan,
    Hunan,
    Jiangxi,
    Shanghai,
    Hangzhou,
    Ningbo,
    Northern,
    #[default]
    Other, // For any other type not listed
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Default)]
pub struct SocialMedia {
    platform: SocialMediaPlatform,
    handle: String,      // e.g., @starbucks
    url: Option<String>, // e.g., https://www.instagram.com/starbucks
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum SocialMediaPlatform {
    Facebook,
    Twitter,
    Instagram,
    LinkedIn,
    TikTok,
    WeChat,
    Weibo,
    Bilibili,
    RedNote,
    Reddit,
    Discord,
    Bluesky,
    WhatsApp,
    Telegram,
    Other(String),
}

impl Default for SocialMediaPlatform {
    fn default() -> Self {
        SocialMediaPlatform::Other("Unknown".to_string())
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Copy, Default)]
#[serde(rename_all = "kebab-case")]
pub enum MerchantStyle {
    #[default]
    Store,
    Kiosk,
    PopUp,
    FoodTruck,
    Room,
}

impl Service for Merchant {
    fn get_id(&self) -> String {
        self.id.to_hex()
    }

    fn get_name(&self) -> String {
        self.name.clone()
    }

    fn set_name(&mut self, name: String) {
        self.name = name;
    }

    fn get_description(&self) -> Option<String> {
        self.description.clone()
    }

    fn set_description(&mut self, description: Option<String>) {
        self.description = description;
    }

    fn get_collection_name() -> &'static str {
        "merchants"
    }

    fn require_unique_name() -> bool {
        false
    }
}

File: mobile/src/lib/structure/merchant.ts
import type {
  MerchantType,
  FoodCuisine,
  ChineseFoodCuisine,
  FoodType,
} from '@/schema'

/**
 * Formats a MerchantType for English display
 * @param type The merchant type to format
 * @returns A human-readable English string representation
 */
export function formatMerchantType(type: MerchantType): string {
  if (typeof type === 'string') {
    return formatSimpleType(type)
  }

  if ('food' in type) {
    return formatFoodType(type.food)
  }

  if ('electronics' in type) {
    return formatElectronicsType(type.electronics)
  }

  if ('clothing' in type) {
    return formatClothingType(type.clothing)
  }

  return 'Unknown'
}

/**
 * Formats simple string merchant types
 */
function formatSimpleType(type: string): string {
  const typeMap: Record<string, string> = {
    supermarket: 'Supermarket',
    health: 'Health & Medical',
    entertainment: 'Entertainment',
    service: 'Service',
    room: 'Room',
    other: 'Other',
  }

  return typeMap[type] || type.charAt(0).toUpperCase() + type.slice(1)
}

/**
 * Formats food merchant types with cuisine and type details
 */
function formatFoodType(food: {
  cuisine: FoodCuisine | null
  type: FoodType
}): string {
  const foodTypeStr = formatFoodTypeOnly(food.type)
  const cuisineStr = food.cuisine ? formatCuisine(food.cuisine) : null

  if (cuisineStr && foodTypeStr.toLowerCase() !== 'restaurant') {
    return `${cuisineStr} ${foodTypeStr}`
  } else if (cuisineStr) {
    return `${cuisineStr} Restaurant`
  } else {
    return foodTypeStr
  }
}

/**
 * Formats the food type portion only
 */
function formatFoodTypeOnly(foodType: FoodType): string {
  if (typeof foodType === 'string') {
    const typeMap: Record<string, string> = {
      cafe: 'Cafe',
      fastFood: 'Fast Food',
      bakery: 'Bakery',
      bar: 'Bar',
      other: 'Restaurant',
    }
    return typeMap[foodType] || 'Restaurant'
  }

  if ('restaurant' in foodType) {
    return 'Restaurant'
  }

  return 'Restaurant'
}

/**
 * Formats cuisine types with specific details
 */
function formatCuisine(cuisine: FoodCuisine): string {
  if (typeof cuisine === 'string') {
    const cuisineMap: Record<string, string> = {
      italian: 'Italian',
      indian: 'Indian',
      american: 'American',
      japanese: 'Japanese',
      korean: 'Korean',
      french: 'French',
      thai: 'Thai',
      mexican: 'Mexican',
      mediterranean: 'Mediterranean',
      spanish: 'Spanish',
      vietnamese: 'Vietnamese',
      middleEastern: 'Middle Eastern',
      african: 'African',
    }
    return (
      cuisineMap[cuisine] || cuisine.charAt(0).toUpperCase() + cuisine.slice(1)
    )
  }

  if ('chinese' in cuisine) {
    return formatChineseCuisine(cuisine.chinese)
  }

  if ('other' in cuisine) {
    return cuisine.other
  }

  return 'International'
}

/**
 * Formats Chinese cuisine with regional specifics
 */
function formatChineseCuisine(chinese: {
  cuisine: ChineseFoodCuisine
  specific: string | null
}): string {
  const regionMap: Record<ChineseFoodCuisine, string> = {
    cantonese: 'Cantonese',
    sichuan: 'Sichuan',
    hunan: 'Hunan',
    jiangxi: 'Jiangxi',
    shanghai: 'Shanghai',
    hangzhou: 'Hangzhou',
    ningbo: 'Ningbo',
    northern: 'Northern Chinese',
    other: 'Chinese',
  }

  const regionStr = regionMap[chinese.cuisine]

  if (chinese.specific) {
    return `${regionStr} (${chinese.specific})`
  }

  return `${regionStr} Chinese`
}

/**
 * Formats electronics merchant types with category details
 */
function formatElectronicsType(electronics: {
  is_mobile: boolean
  is_computer: boolean
  is_accessories: boolean
}): string {
  const categories: string[] = []

  if (electronics.is_mobile) categories.push('Mobile Devices')
  if (electronics.is_computer) categories.push('Computers')
  if (electronics.is_accessories) categories.push('Accessories')

  if (categories.length === 0) {
    return 'Electronics Store'
  } else if (categories.length === 1) {
    return `${categories[0]} Store`
  } else if (categories.length === 2) {
    return `${categories.join(' & ')} Store`
  } else {
    return 'Electronics Store'
  }
}

/**
 * Formats clothing merchant types with target demographic details
 */
function formatClothingType(clothing: {
  is_menswear: boolean
  is_womenswear: boolean
  is_childrenswear: boolean
}): string {
  const demographics: string[] = []

  if (clothing.is_menswear) demographics.push("Men's")
  if (clothing.is_womenswear) demographics.push("Women's")
  if (clothing.is_childrenswear) demographics.push("Children's")

  if (demographics.length === 0) {
    return 'Clothing Store'
  } else if (demographics.length === 1) {
    return `${demographics[0]} Clothing`
  } else if (demographics.length === 2) {
    return `${demographics.join(' & ')} Clothing`
  } else {
    return 'Family Clothing Store'
  }
}

/**
 * Gets a short category label for the merchant type (useful for tags, filters)
 * @param type The merchant type to categorize
 * @returns A short category string
 */
export function getMerchantCategory(type: MerchantType): string {
  if (typeof type === 'string') {
    return type.charAt(0).toUpperCase() + type.slice(1)
  }

  if ('food' in type) {
    return 'Food & Dining'
  }

  if ('electronics' in type) {
    return 'Electronics'
  }

  if ('clothing' in type) {
    return 'Fashion'
  }

  return 'Other'
}

/**
 * Gets detailed type information as an object (useful for filtering/search)
 * @param type The merchant type to analyze
 * @returns An object with detailed type information
 */
export function getMerchantTypeDetails(type: MerchantType): {
  category: string
  subcategory?: string
  details?: string[]
  displayName: string
} {
  const displayName = formatMerchantType(type)

  if (typeof type === 'string') {
    return {
      category: formatSimpleType(type),
      displayName,
    }
  }

  if ('food' in type) {
    const details: string[] = []
    if (type.food.cuisine) {
      details.push(`Cuisine: ${formatCuisine(type.food.cuisine)}`)
    }
    details.push(`Type: ${formatFoodTypeOnly(type.food.type)}`)

    return {
      category: 'Food & Dining',
      subcategory: formatFoodTypeOnly(type.food.type),
      details,
      displayName,
    }
  }

  if ('electronics' in type) {
    const details: string[] = []
    if (type.electronics.is_mobile) details.push('Mobile Devices')
    if (type.electronics.is_computer) details.push('Computers')
    if (type.electronics.is_accessories) details.push('Accessories')

    return {
      category: 'Electronics',
      subcategory: 'Electronics Store',
      details:
        details.length > 0 ? [`Categories: ${details.join(', ')}`] : undefined,
      displayName,
    }
  }

  if ('clothing' in type) {
    const details: string[] = []
    if (type.clothing.is_menswear) details.push("Men's")
    if (type.clothing.is_womenswear) details.push("Women's")
    if (type.clothing.is_childrenswear) details.push("Children's")

    return {
      category: 'Fashion',
      subcategory: 'Clothing Store',
      details:
        details.length > 0 ? [`Target: ${details.join(', ')}`] : undefined,
      displayName,
    }
  }

  return {
    category: 'Other',
    displayName,
  }
}
